[crowdstrike] Collector crowdstrike - chunk 1, 2 (#11)

savacano28 opened the pull request:
<!--
Thank you very much for your pull request to the OpenBAS project! We as a community driven project depend on support and contributions like this!

Thus already a BIG THANK YOU upfront to you for choosing to help with your PR.
-->

### Proposed changes

* Add new collector for Crowdstrike
* Enable python tests in CI

### Related issues

* Fix #11 

### Checklist

<!--
Please submit the source code in a way, where you could honestly say `This code is finished`.
If you feel that there are possibilities for improving the code quality, please do so.
By doing this, you are actively helping us to improve the quality of the entire OpenBAS project.
-->

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

<!-- For completed items, change [ ] to [x]. -->

### Further comments

The chosen approach is to decouple the implementation details of the specific Crowdstrike API (currently: `alerts`) from the broader logic, so as to future-proof the collector if we need to switch to another API.

savacano28 committed:
[crowdstrike] add collector crowdstrike

savacano28 committed:
[crowdstrike] refact

savacano28 committed:
[crowdstrike] refact

savacano28 committed:
[crowdstrike] refact

savacano28 committed:
[crowdstrike] wip fecth expectations from openbas

savacano28 committed:
[crowdstrike] clean

savacano28 committed:
[crowdstrike] clean

savacano28 committed:
[crowdstrike] refact call cs endpoints

savacano28 committed:
[crowdstrike] add call detects scope and refact

savacano28 committed:
[crowdstrike] add platform

savacano28 committed:
[crowdstrike] add platform

savacano28 committed:
[crowdstrike] lint

savacano28 committed:
[crowdstrike] add security platform in config

antoinemzs committed:
Merge branch 'main' into issue/11-match-expectations-with-cs

antoinemzs committed:
overhaul with new features in client

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
fix modules and test in ci

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
black

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
enable tests for real

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
linter and formatter fixes

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
add more tests

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
complete test of alert strategy

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
implement determining prevention

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
black + isort

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
fix linting issues

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
start integration tests

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
finalise tests for crowdstrike collector

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
linting/formatting

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
prop up readme

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs commented on crowdstrike/crowdstrike/openbas_crowdstrike.py:
```diff
@@ -0,0 +1,234 @@
+                        result = "Detected"
+                    elif expectation.get("inject_expectation_type") == "PREVENTION":
+                        success_or_failure = self.strategy.is_prevented(alert)
+                        result = "Prevented" if success_or_failure else "Not prevented"
```
Unfortunately this is still a bit stringly typed

antoinemzs commented on crowdstrike/crowdstrike/openbas_crowdstrike.py:
```diff
@@ -0,0 +1,234 @@
+    def _process(self):
+        """Fetch and match expectations with data from cs"""
+        now = datetime.now(pytz.UTC)
+        start_time = now - timedelta(minutes=45)
```
This is the same hardcoded timeframe as for the Defender collector. Not a huge change to make it a config, but let's see if this is an issue? Crowdstrike is usually very quick to collect data from endpoints.

antoinemzs commented on crowdstrike/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+pyobas @ git+https://github.com/OpenBAS-Platform/client-python@release/1.10.0
```
Possibly I should align this with the upcoming changes related to the prerelease deployment target.

RomuDeuxfois commented on crowdstrike/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM python:3.12-alpine
+ENV CONNECTOR_TYPE=EXTERNAL_IMPORT
```
**Question**
Why we have this ?

antoinemzs replied:
It's a default configuration for the collector, but it looks like a c/c mistake (aka copypasta) from possibly an existing file from opencti (hence "connector" and not collector).

This sets a defaults that overrides any default set in code, and I don't think it's necessary at all (rather, it would cause surprises)

RomuDeuxfois commented on crowdstrike/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+ENV CONNECTOR_TYPE=EXTERNAL_IMPORT
+
+# Copy the connector
+ADD crowdstrike /opt/openbas-collector-crowdstrike/crowdstrike
+WORKDIR /opt/openbas-collector-crowdstrike
+ADD requirements.txt .
+
+# Install Python modules
+# hadolint ignore=DL3003
+RUN apk update && apk upgrade && \
+    apk --no-cache add git build-base libmagic libffi-dev libxslt libxslt-dev libxml2 libxml2-dev && \
+    pip install --no-cache-dir -r requirements.txt && \
+    apk del git build-base && \
+    rm -rf /var/cache/apk/*
+
+CMD ["python3", "-m", "crowdstrike.openbas_crowdstrike"]
```
**thought**
i don't know if it's a good idea to try homogenize our Dockerfile.
We have not the same to do it each time.

antoinemzs replied:
We could definitely build and provide a base image with packages preinstalled. However I think we can optimise the images better by switching to a multi-stage build process. We could have a discussion about this.

RomuDeuxfois commented on crowdstrike/README.md:
````diff
@@ -0,0 +1,37 @@
+pip install -r requirements.txt .
+```
+
+## Configuration
+
+The connector can be configured with the following variables:
+
````
**Question**
Do we need any set up on crowdstrike for API Call ? (permissions for example)

antoinemzs replied:
I'll look for this. For now I suspect we have been using very privileged credentials for development.

RomuDeuxfois commented on crowdstrike/crowdstrike/openbas_crowdstrike.py:
```diff
@@ -0,0 +1,234 @@
+
+        valid_expectations = []
+
+        for expectation in expectations:
```
Can we add a if to don't go further if we have any(expectations) ?
Or it's not necessary ?

antoinemzs replied:
for loops don't iterate at all on empty lists

RomuDeuxfois commented on crowdstrike/crowdstrike/openbas_crowdstrike.py:
```diff
@@ -0,0 +1,234 @@
+                    expectation["inject_expectation_id"],
+                    {
+                        "collector_id": self.config.get_conf("collector_id"),
+                        "result": (
+                            "Not Detected"
+                            if expectation["inject_expectation_type"] == "DETECTION"
+                            else "Not Prevented"
+                        ),
```
**Suggestion**

Should be great to have an utils for this (use in MDE too)

antoinemzs replied:
Agreed, however to achieve this we need to create common components and it's probably outside of the scope of this feature by now... I created #75 to consider this.

As it's not a big change, it could be sneaked into another feature some time.

RomuDeuxfois commented on crowdstrike/crowdstrike/pattern_disposition.py:
```diff
@@ -0,0 +1,53 @@
+# took in a bitmask field. Thankfully, the mask components
+# are documented so we can create this utility.
+# Reference:https://falcon.us-2.crowdstrike.com/documentation/page/e3ce0b24/events-data-dictionary#ModuleBlockedEventWithPatternId
+PREVENT_KILL_PROCESS = 16
```
**praise**
Good shot

antoinemzs replied:
🎯

RomuDeuxfois commented on crowdstrike/config.yml.sample:
```diff
@@ -0,0 +1,16 @@
+openbas:
```
Move it to crowdstrike repo

antoinemzs committed:
Add pycharm configuration info

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
bit more info about tests in dev mode in readme

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
ignore hostname in signatures

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
more simpler + black/isort

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
some review comment fixes

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>

antoinemzs committed:
align requirements.txt to branch practice

note this will not provide the correct pyobas version for a lot of cases
in development work whenever there is a dependency on an unreleased
version of pyobas.

To work around this, individual developers should manually checkout and install the
unreleased pyobas in their environment after installing all other requirements.

Signed-off-by: Antoine MAZEAS <antoine.mazeas@filigran.io>
